Xfem crack nucleation in bulk (#27055)
idaholab/moose: modules/xfem/doc/content/source/userobjects/MeshCut2DRankTwoTensorNucleation.md

If the scalar exceeds a threshold specified by [!param](/UserObjects/MeshCut2DRankTwoTensorNucleation/nucleation_threshold), a line segment with a length specified by [!param](/UserObjects/MeshCut2DRankTwoTensorNucleation/nucleation_length) will be inserted
into the [MeshCut2DFractureUserObject.md] cutter mesh.
Cracks will nucleate on the domain interior or boundary.  Crack nucleation can be limited to elements on boundaries specified by [!param](/UserObjects/MeshCut2DRankTwoTensorNucleation/initiate_on_boundary).
Once the nucleation criterion is reached, a line segment of the specified length is inserted into the cutter mesh,
centered on the element centroid it nucleates from.
The direction of the nucleated crack is normal to the direction returned by the `RankTwoTensor` scalar.
For example, `MaxInPlanePrincipal` returns the direction of the maximum in-plane principal component and the crack direction is normal to this.

> It is an error for [!param](/UserObjects/MeshCut2DRankTwoTensorNucleation/nucleation_length) to be smaller than element the crack nucleates in .

Reviewer: ```suggestion
It is an error for [!param](/UserObjects/MeshCut2DRankTwoTensorNucleation/nucleation_length) to be smaller than the length of the element the crack nucleates in .
```

> It is an error for [!param](/UserObjects/MeshCut2DRankTwoTensorNucleation/nucleation_length) to be smaller than the length of the element the crack nucleates in .

Reviewer: ```suggestion
It is an error for [!param](/UserObjects/MeshCut2DRankTwoTensorNucleation/nucleation_length) to be smaller than the length of the element the crack nucleates in, unless the user sets `nucleate_across_full_element=true`.
```

A crack will only be nucleated if it is at least a distance specified by [!param](/UserObjects/MeshCut2DRankTwoTensorNucleation/nucleation_radius) away from existing or nucleated cracks.

> If multiple cracks nucleate in the same xfem update and are within the specified `nucleation_radius`, the crack nucleated from the element with the lowest id will be retained and no other cracks within the `nucleation_radius` will be nucleated.  The `nucleation_radius` is a good way to limit the number of cracks that nucleate within a region, which can help limit problems related to excessive numbers of XFEM cuts in a single element.

Reviewer: Does it apply only if they nucleate during the same xfem update? If I nucleate a crack at one time step, and then another one wants to nucleate close to it later on, is it allowed even if `nucleation_radius` is set?

> If multiple cracks nucleate in the same xfem update and are within the specified `nucleation_radius`, the crack nucleated from the element with the lowest id will be retained and no other cracks within the `nucleation_radius` will be nucleated.  The `nucleation_radius` is a good way to limit the number of cracks that nucleate within a region which can lead to XFEM errors related to cutting element fragments where an element contains more than one XFEM cut.

Reviewer: ```suggestion
If multiple cracks nucleate in the same xfem update and are within the specified `nucleation_radius`, the crack nucleated from the element with the lowest id will be retained and no other cracks within the `nucleation_radius` will be nucleated.  The `nucleation_radius` is a good way to limit the number of cracks that nucleate within a region, which can help limit problems related to excessive numbers of XFEM cuts in a single element.
```
